src/orquestra/sdk/_base/_api/_wf_run.py

```diff
@@ -431,11 +433,73 @@ def get_logs(self) -> WorkflowLogs:
         """
         return self._runtime.get_workflow_logs(wf_run_id=self.run_id)
 
-    # TODO: ORQSDK-617 add filtering ability for the users
-    def get_tasks(self) -> t.Set[TaskRun]:
-        wf_run_model = self.get_status_model()
+    def get_tasks(
+        self,
+        state: t.Optional[t.Union[State, t.List[State]]] = None,
+        function_name: t.Optional[str] = None,
+        task_run_id: t.Optional[t.Union[str, TaskRunId]] = None,
+        task_invocation_id: t.Optional[t.Union[str, ir.TaskInvocationId]] = None,
+    ) -> t.Set[TaskRun]:
+        """
+        Returns TaskRun representations of the tasks executed as part of this workflow.
 
-        return {
+        Args:
+            state: If specified, only tasks with matching states will be returned.
+            function_name: A function name, or regex string matching the desired
+                function name(s). If specified, only tasks with matching function names
+                will be returned.
+            task_run_id: A task run ID, or regex string matching the desired task run
+                ID(s). If specified, only tasks with matching task run IDs will be
+                returned.
+            task_invocation_id: A task invocation ID, or regex string matching the
+                desired task invocation ID(s). If specified, only tasks with matching
+                task invocation IDs will be returned.
+
+        Returns:
+            An iterable of TaskRuns
+        """
+
+        def matches_model_filters(
+            task_run_model: TaskRunModel,
+            state: t.Optional[t.Union[State, t.List[State]]] = None,
+            task_run_id: t.Optional[t.Union[str, TaskRunId]] = None,
+            task_invocation_id: t.Optional[t.Union[str, ir.TaskInvocationId]] = None,
+        ) -> bool:
+            """
+            Filters that can be applied to orquestra.sdk.schema.workflow_run.TaskRun
+            """
+            if state:
+                states: t.List[State]
+                if isinstance(state, State):
+                    states = [state]
+                else:
+                    states = state
+                if task_run_model.status.state not in states:
+                    return False
+
+            if task_run_id and not re.match(task_run_id, task_run_model.id):
+                return False
+
+            if task_invocation_id and not re.match(
+                task_invocation_id, task_run_model.invocation_id
+            ):
+                return False
+
+            return True
+
+        def matches_run_filters(
+            task_run: TaskRun,
+            task_fn_name: t.Optional[str] = None,
+        ) -> bool:
+            """
+            Filters that can applied to orquestra.sdk._base._api._task_run.TaskRun.
+            """
+            if task_fn_name and not re.match(task_fn_name, task_run.fn_name):
+                return False
+            return True
+
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** ‏Can we pull these out for easier unit testing?

Reply:
We can, and as of [98e856a](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/188/commits/98e856a81c6cb5bc7b99c6f9f10169ff89e13aa3) we have done :)